saasherder/saasherder.py

```diff
@@ -129,14 +130,39 @@ def apply_filter(self, template_filter, data):
 
         return yaml.dump(data_obj, encoding='utf-8', default_flow_style=False)
 
-    def apply_saasherder_annotations(self, data, service_name):
+    def apply_saasherder_labels_and_annotations(self, data, service_name, saas_repo_url):
         data_obj = yaml.safe_load(data)
+        data_sha256sum = self.calculate_sha256sum_label(data)
+        labels_selector = 'saasherder.data-sha256sum notin (%s)' % data_sha256sum
+        if saas_repo_url:
+            saas_repo_url_hash = self.calculate_sha256sum_label(saas_repo_url)
+        
         for obj in data_obj.get("items", []):
-            obj['metadata'].setdefault('annotations', {})
-            annotations = obj['metadata']['annotations']
-            annotations['saasherder.service'] = "%s/%s" % (self.config.current(), service_name)
-
-        return yaml.safe_dump(data_obj, encoding='utf-8', default_flow_style=False)
+            obj['metadata'].setdefault('labels', {})
+            labels = obj['metadata']['labels']
+            labels['saasherder.context'] = self.config.current()
+            labels['saasherder.service'] = service_name
+            labels['saasherder.data-sha256sum'] = data_sha256sum[:63]
+            labels_selector = \
+                '%s, saasherder.context in (%s), saasherder.service in (%s)' \
+                % (labels_selector, self.config.current(), service_name)
+
+            if saas_repo_url:
+                labels['saasherder.saas-repo-url-sha256sum'] = saas_repo_url_hash
+                labels_selector = \
+                    "%s, saasherder.saas-repo-url-sha256sum in (%s)" \
+                    % (labels_selector, saas_repo_url_hash)
+                # add annotation for human readability
+                obj['metadata'].setdefault('annotations', {})
+                annotations = obj['metadata']['annotations']
+                annotations['saasherder.saas-repo-url'] = saas_repo_url
+
+        return yaml.safe_dump(data_obj, encoding='utf-8', default_flow_style=False), labels_selector
+
+    def calculate_sha256sum_label(self, data):
+        m = hashlib.sha256()
+        m.update(data)
```

Comment:
I think you can just initialize it with data: `return hashlib.sha256(data).hexdigest()[:10]`

Reply:
thanks!